### PR TITLE
Add functionality to test if the host CPU supports native SIMD instructions

### PR DESCRIPTION
**Pull request by ldematte**
This commit adds the ability to check if the host CPU supports the instruction set we use in our native vector code.
This is to avoid having the process terminated (badly) because we tried to execute a non-supported instruction; better know this early, and avoid to take the native code path entirely.

The `int vec_caps()` function, implemented in the native code (C) of the library, returns 0 if the processor does not support (i.e. cannot run) the instructions used to implement the other functions.

On ARM, we use the NEON vector instruction set (https://github.com/elastic/elasticsearch/pull/106133):
- On MacOS, `vec_caps()` is trivial and returns `1`; for M-series CPUs, we can safely assume NEON is always present.
- On Linux systems, we use the mechanism explained in the [ARM development guide](https://documentation-service.arm.com/static/63299276e68c6809a6b41308): call [`getauxval`](https://www.man7.org/linux/man-pages/man3/getauxval.3.html) with `AT_HWCAP`, and test for the `HWCAP_NEON` bit (4096).

I've tested `getauxval` on various ARM systems and emulators (including Graviton on AWS); the NEON instruction set seems to be widely available - I was not able to find a processor supporting `armv8` and _not_ supporting NEON; nevertheless, I think that adding this functionality is good for 3 reasons:
- it is better to err on the safe side
- we would eventually need this for x64 (where support for vector instruction sets is way more sparse)
- we can expand it for future optimizations

The last point: since `vec_caps()` returns an integer, we may want to return different values (e.g. 2) to indicate some "advanced support", and dynamically bind to a different flavor of the function (e.g. `dot8s_2`) that is implemented with a more performant but less diffused instruction set (e.g. AVX512, armv9, etc.)

### CODE REVIEW

File: libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkVectorLibrary.java [hunk 1]
```diff
@@ -23,142 +24,166 @@
...
 
-    static final MethodHandle dot8s$mh = downcallHandle("dot8s", FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS, JAVA_INT));
-    static final MethodHandle sqr8s$mh = downcallHandle("sqr8s", FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS, JAVA_INT));
+    private static final class JdkVectorSimilarityFunctions implements VectorSimilarityFunctions {
 
-    // Stride of the native implementation - consumes this number of bytes per loop invocation.
-    // There must be at least this number of bytes/elements available when going native
-    static final int DOT_STRIDE = 32;
-    static final int SQR_STRIDE = 16;
+        static final MethodHandle dot8stride$mh = downcallHandle("dot8s_stride", FunctionDescriptor.of(JAVA_INT));
```
Comment (ldematte) [inline libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkVectorLibrary.java:54]: @ChrisHegarty GH made a bit of a mess with the diff, but essentially code from here to the end is the same, just indendeted (encapsulated in the private inner class)

File: libs/native/src/main/java/org/elasticsearch/nativeaccess/VectorSimilarityFunctions.java [hunk 2]
```diff
@@ -8,44 +8,17 @@
...
-     *
-     * <p> The type of the method handle will have {@code int} as return type, The type of
-     * its first and second arguments will be {@code MemorySegment}, whose contents is the
-     * vector data bytes. The third argument is the length of the vector data.
-     */
-    public MethodHandle dotProductHandle() {
-        return vectorLibrary.dotProductHandle();
-    }
+public interface VectorSimilarityFunctions {
+    MethodHandle dotProductHandle();
```
Comment (ChrisHegarty) [inline libs/native/src/main/java/org/elasticsearch/nativeaccess/VectorSimilarityFunctions.java:21]: Can we please retain the javadoc for each of these two methods.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by ldematte** [comment]
@elasticmachine update branch

**Review by ChrisHegarty — APPROVED** [review]

**Comment by ldematte** [comment]
@elasticmachine update branch

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by ldematte** [comment]
@elasticmachine update branch

**Pull request merged**